enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,441 @@
+Routes, the user would have to provide the Certificate CR reference.
+
+This also adds coupling with external third-party API(s) and limits future enhancement to be
+a more generic solution for integrating with third-party certificate management systems.
+
+##### Alternative to new status condition types
+
+An alternative to introducing 3 new condition types, would be to use `metadata.managedFields`
+on Route to denote to the user which fields are managed by the secret-injector controller and
+server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate`
+and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.
+
+This although helps in avoiding addition of new condition types will result in a few inconsistencies,
+
+- When transitioning from using managed certificates to manually managing custom certificates,
+  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update
+  operation done by the user is only on the `.metadata.annotations`.
+- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.
+
+### Implementation Details/Notes/Constraints [optional]
+
+The new controller (secret-injector) will be responsible for watching routes and processing those
+that have the valid annotation for `tls-secret-name`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+Since secrets are a related resource to the secret-injector controller, the new controller
+will also set up watches and re-sync routes in the same namespace as that of the secret (similar
+to ingress-to-route controller).
+
+The new controller will watch all routes and process all routes in order to update
+`.status.certificate`. The ones with the annotation will have additional
+logic to update `.spec.tls.certificate` and `.spec.tls.key` based on the following pre-conditions,
+
+- Annotation is present and valid
+- Secret mentioned in the annotation exists and is valid
+
+In scenarios where the user deletes the annotation, it is the controller's
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.
+if the status denotes `ManagedCertificate` and since the required annotation is
+not present this results in the `.spec.tls.certificate` and `.spec.tls.key` being cleared
+out and the `.status` updated to `DefaultCertificate`.
+
+If the user deletes the secret associated with the Route without deleting the
+`route.openshift.io/tls-secret-name` annotation, the secret-injector controller will
+clear out `.spec.tls.certificate` and `.spec.tls.key` since the following pre-conditions
+are met,
+
+- Annotation is present and valid
+- Secret mentioned in the annotation doesn't exist
+- Route `.status.certificate` has `ManagedCertificate`=`True`
```

Comment:
Shouldn't this be the latest condition to be precise?

Reply:
Addressed.